cardano-api/src/Cardano/Api/IPC/Monad.hs

```diff
@@ -17,54 +22,148 @@ import Control.Monad.Trans.Cont
+      -> TMVar ChainTip
+      -> ChainSyncClient (BlockInMode mode) ChainPoint ChainTip IO ()
+    chainSyncGetCurrentTip
+        waitDone  -- ^ An STM expression that only returns when all protocols are complete.
+                  -- Protocols must wait until 'waitDone' returns because premature exit will
+                  -- cause other incomplete protocols to abort which may lead to deadlock.
+        tipVar =
+      ChainSyncClient $ pure clientStIdle
+      where
+        clientStIdle :: Net.Sync.ClientStIdle (BlockInMode mode) ChainPoint ChainTip IO ()
+        clientStIdle =
+          Net.Sync.SendMsgRequestNext clientStNext (pure clientStNext)
+
+        clientStNext :: Net.Sync.ClientStNext (BlockInMode mode) ChainPoint ChainTip IO ()
+        clientStNext = Net.Sync.ClientStNext
+          { Net.Sync.recvMsgRollForward = \_block tip -> ChainSyncClient $ do
+              void . atomically $ putTMVar tipVar tip
+              void $ atomically waitDone -- Wait for all protocols to complete before exiting.
+              pure $ Net.Sync.SendMsgDone ()
+          , Net.Sync.recvMsgRollBackward = \_point tip -> ChainSyncClient $ do
+              void . atomically $ putTMVar tipVar tip
+              void $ atomically waitDone -- Wait for all protocols to complete before exiting.
+              pure $ Net.Sync.SendMsgDone ()
           }
-      else do
-        atomically $ putTMVar resultVar' Nothing
-        void $ atomically waitDone
+
+-- | Use 'queryExpr' in a do block to construct monadic local state queries.
+setupLocalStateQueryExpr ::
+     STM x
+  -> Maybe ChainPoint
+  -> TMVar (Either Net.Query.AcquireFailure a)
+  -> LocalStateQueryExpr (BlockInMode mode) ChainPoint (QueryInMode mode) () IO a
+  -> Net.Query.LocalStateQueryClient (BlockInMode mode) ChainPoint (QueryInMode mode) IO ()
+setupLocalStateQueryExpr
+    waitDone  -- ^ An STM expression that only returns when all protocols are complete.
+              -- Protocols must wait until 'waitDone' returns because premature exit will
+              -- cause other incomplete protocols to abort which may lead to deadlock.
+    mPointVar'
+    resultVar'
+    f =
+  LocalStateQueryClient . pure . Net.Query.SendMsgAcquire mPointVar' $
+    Net.Query.ClientStAcquiring
+    { Net.Query.recvMsgAcquired = runContT (runLocalStateQueryExpr f) $ \result -> do
+        atomically $ putTMVar resultVar' (Right result)
+        void $ atomically waitDone -- Wait for all protocols to complete before exiting.
+        pure $ Net.Query.SendMsgRelease $ pure $ Net.Query.SendMsgDone ()
+
+    , Net.Query.recvMsgFailure = \failure -> do
+        atomically $ putTMVar resultVar' (Left failure)
+        void $ atomically waitDone -- Wait for all protocols to complete before exiting.
         pure $ Net.Query.SendMsgDone ()
+    }
+
+-- | Use 'queryExpr' in a do block to construct monadic local state queries.
+queryExpr :: query a -> LocalStateQueryExpr block point query r IO a
```

Comment:
`query a` -> `QueryInMode mode result` ? We should be able to do that.

Reply:
Done!